### PR TITLE
Stellar dotnet sdk

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Lykke.Service.Stellar.Api.Services/Horizon/HorizonService.cs [hunk 1]
```diff
@@ -135,94 +145,117 @@ public async Task<bool> AccountExists(string address)
...
-            writer.Write(Network.CurrentNetworkId);
-
-            // Envelope Type - 4 bytes
-            EnvelopeType.Encode(writer, EnvelopeType.Create(EnvelopeType.EnvelopeTypeEnum.ENVELOPE_TYPE_TX));
-
-            // Transaction XDR bytes
-            var txWriter = new ByteWriter();
-            StellarBase.Generated.Transaction.Encode(txWriter, tx);
-            writer.Write(txWriter.ToArray());
+            throw new NotImplementedException();
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Stellar.Api.Services/Horizon/HorizonService.cs:230]: Is this method needed if it's not implemented?

File: src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs [hunk 2]
```diff
@@ -213,34 +210,36 @@ private static string GetErrorMessage(Exception ex)
         {
             var errorMessage = ex.Message;
             // ReSharper disable once InvertIf
-            if (ex is BadRequestException badRequest)
+            if (ex is HttpResponseException badRequest)
             {
-                var resultCodes = JsonConvert.SerializeObject(badRequest.ErrorDetails.Extras.ResultCodes);
-                errorMessage += $". ResultCodes={resultCodes}";
+                //TODO:
+                //var resultCodes = JsonConvert.SerializeObject(badRequest.ErrorDetails.Extras.ResultCodes);
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs:216]: Haven't we missed this?

File: src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs [hunk 3]
```diff
@@ -213,34 +210,36 @@ private static string GetErrorMessage(Exception ex)
...
-            {
-                return TxExecutionError.NotEnoughBalance;
-            }
-
-            if (transactionDetail == "tx_too_late" ||
-                transactionDetail == "tx_bad_seq")
-            {
-                return TxExecutionError.BuildingShouldBeRepeated;
-            }
+            //TODO:
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs:224]: Is there a way to process error and return correct code?

File: src/Lykke.Service.Stellar.Api.Services/Horizon/HorizonService.cs (hunk 1 above)
Comment (OttoVT) [reply src/Lykke.Service.Stellar.Api.Services/Horizon/HorizonService.cs:230]: fixed

File: src/Lykke.Service.Stellar.Api.Core/Exceptions/BadRequestHorizonApiException.cs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+using System.Linq;
+
+namespace Lykke.Service.Stellar.Api.Core.Exceptions
+{
+    public class BadRequestHorizonApiException : Exception
+    {
+        public BadRequestHorizonApiException(string message, IReadOnlyCollection<string> errorCodes)
+            : base(message)
+        {
+            ErrorCodes = errorCodes?.ToArray() ?? new String[]{};
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Stellar.Api.Core/Exceptions/BadRequestHorizonApiException.cs:12]: Don't need to change here, but `Array.Empty<string>()` for the future

File: src/Lykke.Service.Stellar.Api.Services/Horizon/HorizonService.cs [hunk 5]
```diff
@@ -49,12 +49,18 @@ public async Task<string> SubmitTransactionAsync(string signedTx)
             var acc = await _server.Accounts.Account(transaction.SourceAccount.AccountId);
             var tx = await _server.SubmitTransaction(transaction);
 
-
-            if (tx == null || string.IsNullOrEmpty(tx.Hash))
+            if (string.IsNullOrEmpty(tx?.Hash))
             {
+                if (!tx.Result.IsSuccess)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Stellar.Api.Services/Horizon/HorizonService.cs:54]: It looks like NRE possible here. Shouldn't we throw HorizonApiException in the case when tx is null and if not, then check IsSuccess?

File: src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs [hunk 6]
```diff
@@ -210,36 +212,37 @@ private static string GetErrorMessage(Exception ex)
...
-            //    return TxExecutionError.BuildingShouldBeRepeated;
-            //}
+            var bre = ex as BadRequestHorizonApiException;
+            var resultCodes = bre?.ErrorCodes;
+            var transactionDetail = bre?.Message;
+
+            if (transactionDetail != null
+                && resultCodes != null
+                && resultCodes.Length > 0
+                && (resultCodes[0].Equals(StellarSdkConstants.OperationUnderfunded)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs:235]: Why do we check exactly first item, isn't `Contains` fee better here?

Comment (OttoVT) [reply src/Lykke.Service.Stellar.Api.Services/Transaction/TransactionService.cs:235]: nope, It is how it worked before, we do not need to fix this

**Pull request merged**